gto-support-db/src/main/java/org/ccci/gto/android/common/db/AbstractDao.java

```diff
@@ -648,6 +615,45 @@ public final Transaction beginTransaction() {
         return newTransaction().beginTransaction();
     }
 
+    @WorkerThread
+    public final <T, X extends Throwable> T inTransaction(@NonNull final Closure<T, X> closure) throws X {
+        return inTransaction(getWritableDatabase(), closure);
+    }
+
+    @WorkerThread
+    protected final <T, X extends Throwable> T inTransaction(@NonNull final SQLiteDatabase db,
+                                                             @NonNull final Closure<T, X> closure) throws X {
+        final Transaction tx = newTransaction(db);
+        try {
+            tx.beginTransaction();
+            final T result = closure.run();
+            tx.setTransactionSuccessful();
+            return result;
+        } finally {
+            tx.endTransaction().recycle();
+        }
+    }
+
+    @WorkerThread
+    public final <T, X extends Throwable> T inNonExclusiveTransaction(@NonNull final Closure<T, X> closure) throws X {
+        return inNonExclusiveTransaction(getWritableDatabase(), closure);
+    }
+
+    @WorkerThread
+    protected final <T, X extends Throwable> T inNonExclusiveTransaction(@NonNull final SQLiteDatabase db,
+                                                                         @NonNull final Closure<T, X> closure)
+            throws X {
+        final Transaction tx = newTransaction(db);
+        try {
+            tx.beginTransactionNonExclusive();
+            final T result = closure.run();
+            tx.setTransactionSuccessful();
+            return result;
```

Comment:
@frett Just had a question about logic since you requiring the caller of this method to handle the Exception.  In side of the try block will the finally run if there is an exception.  This is something that I haven't seen so just want to learn.

Reply:
it's main use is to clean up resources that might leak no matter what happens. In this case it is ending the transaction so that the db handle is back to it's original state and can be re-used.